File: Matrox Imaging Library (MIL) X 23H2 (10.60)/Examples/General/MdispOverlay/C++/MdispOverlay.cpp
/*************************************************************************************/
/*
 * File name: MDispOverlay.cpp
 *
 * Synopsis:  This program shows how to display an image while creating
 *            text and graphics annotations on top of it using MIL
 *            graphic functions and windows GDI drawing under Windows
 *            or cairo drawing under Linux.
 *            If the target system supports grabbing, the annotations
 *            are done on top of a continuous grab.
 *
 * Copyright © Matrox Electronic Systems Ltd., 1992-2023.
 * All Rights Reserved
 */
#include <mil.h>
#if M_MIL_USE_WINDOWS
#include <windows.h>
#else
#include <stdlib.h>
#include <cairo.h>
#endif

/* Target image. */
#define IMAGE_FILE     M_IMAGE_PATH MIL_TEXT("Board.mim")

/* Title for the display window. */
#define WINDOW_TITLE   MIL_TEXT("Custom Title")

/* Overlay drawing function prototype. */
void OverlayDraw(MIL_ID Display);

int MosMain(void)
   {
   MIL_ID MilApplication,            /* Application identifier. */
          MilSystem,                 /* System identifier.      */
          MilDisplay,                /* Display identifier.     */
          MilDigitizer=0,            /* Digitizer identifier.   */
          MilImage;                  /* Image identifier.       */

   /* Allocate defaults */
   MappAllocDefault(M_DEFAULT, &MilApplication, &MilSystem, &MilDisplay, M_NULL, M_NULL);

   /* If the system have a digitizer, use it. */
   if (MsysInquire(MilSystem, M_DIGITIZER_NUM, M_NULL))
      {
      MdigAlloc(MilSystem, M_DEFAULT, MIL_TEXT("M_DEFAULT"), M_DEFAULT, &MilDigitizer);
      MbufAllocColor(MilSystem,
                     MdigInquire(MilDigitizer,M_SIZE_BAND,M_NULL),
                     MdigInquire(MilDigitizer,M_SIZE_X,M_NULL),
                     MdigInquire(MilDigitizer,M_SIZE_Y,M_NULL),
                     8+M_UNSIGNED,
                     M_IMAGE+M_DISP+M_PROC+M_GRAB,
                     &MilImage);
      MbufClear(MilImage, 0);
      }
   else
      {
      /* Restore a static image. */
      MbufRestore(IMAGE_FILE, MilSystem, &MilImage);
      }

   /* Change display window title. */
   MdispControl(MilDisplay, M_TITLE, WINDOW_TITLE);

   /* Display the image buffer. */
   MdispSelect(MilDisplay, MilImage);

   /* Draw text and graphics annotations in the display overlay. */
   OverlayDraw(MilDisplay);

   /* If the system supports it, grab continuously in the displayed image. */
   if (MilDigitizer)
      MdigGrabContinuous(MilDigitizer, MilImage);

   /* Pause to show the image. */
   MosPrintf(MIL_TEXT("\nOVERLAY ANNOTATIONS:\n"));
   MosPrintf(MIL_TEXT("--------------------\n\n"));
   MosPrintf(MIL_TEXT("Displaying an image with overlay annotations.\n"));
   MosPrintf(MIL_TEXT("Press <Enter> to continue.\n\n"));
   MosGetch();

   /* Stop the continuous grab and free digitizer if needed. */
   if (MilDigitizer)
      {
      MdigHalt(MilDigitizer);
      MdigFree(MilDigitizer);

      /* Pause to show the result. */
      MosPrintf(MIL_TEXT("Displaying the last grabbed image.\n"));
      MosPrintf(MIL_TEXT("Press <Enter> to end.\n\n"));
      MosGetch();
      }

   /* Free image. */
   MbufFree(MilImage);

   /* Free default allocations. */
   MappFreeDefault(MilApplication, MilSystem, MilDisplay, M_NULL, M_NULL);

   return 0;
   }


/* --------------------------------------------------------------- */
/* Name:      OverlayDraw
 * Synopsis:  This function draws annotations in the display overlay.
 */
void OverlayDraw(MIL_ID MilDisplay)
   {
   MIL_ID         MilOverlayImage;
   MIL_INT        ImageWidth, ImageHeight;
   int            Count;
   MIL_TEXT_CHAR chText[80];

   /* Prepare overlay buffer. */
   /***************************/

   /* Enable the display of overlay annotations. */
   MdispControl(MilDisplay, M_OVERLAY, M_ENABLE);

   /* Inquire the overlay buffer associated with the display. */
   MdispInquire(MilDisplay, M_OVERLAY_ID, &MilOverlayImage);

   /* Clear the overlay to transparent. */
   MdispControl(MilDisplay, M_OVERLAY_CLEAR, M_DEFAULT);

   /* Disable the overlay display update to accelerate annotations. */
   MdispControl(MilDisplay, M_OVERLAY_SHOW, M_DISABLE);

   /* Inquire overlay size. */
   ImageWidth  = MbufInquire(MilOverlayImage, M_SIZE_X, M_NULL);
   ImageHeight = MbufInquire(MilOverlayImage, M_SIZE_Y, M_NULL);

   /* Draw MIL overlay annotations. */
   /*********************************/

   /* Set the graphic text background to transparent. */
   MgraControl(M_DEFAULT, M_BACKGROUND_MODE, M_TRANSPARENT);

   /* Print a white string in the overlay image buffer. */
   MgraColor(M_DEFAULT, M_COLOR_WHITE);
   MgraText(M_DEFAULT, MilOverlayImage, ImageWidth/9, ImageHeight/5,    
                                             MIL_TEXT(" -------------------- "));
   MgraText(M_DEFAULT, MilOverlayImage, ImageWidth/9, ImageHeight/5+25, 
                                             MIL_TEXT(" - MIL Overlay Text - "));
   MgraText(M_DEFAULT, MilOverlayImage, ImageWidth/9, ImageHeight/5+50, 
                                             MIL_TEXT(" -------------------- "));

   /* Print a green string in the overlay image buffer. */
   MgraColor(M_DEFAULT, M_COLOR_GREEN);
   MgraText(M_DEFAULT, MilOverlayImage, ImageWidth*11/18, ImageHeight/5,    
                                             MIL_TEXT(" ---------------------"));
   MgraText(M_DEFAULT, MilOverlayImage, ImageWidth*11/18, ImageHeight/5+25, 
                                             MIL_TEXT(" - MIL Overlay Text - "));
   MgraText(M_DEFAULT, MilOverlayImage, ImageWidth*11/18, ImageHeight/5+50, 
                                             MIL_TEXT(" ---------------------"));

   /* Re-enable the overlay display after all annotations are done. */
   MdispControl(MilDisplay, M_OVERLAY_SHOW, M_ENABLE);


   /* Draw GDI/Cairo color overlay annotation */
   /*************************************/

   /* Disable error printing for the inquire might not be supported */
   MappControl(M_DEFAULT, M_ERROR, M_PRINT_DISABLE);

#if M_MIL_USE_WINDOWS
   HDC            hCustomDC;
   HGDIOBJ        hpen, hpenOld;
   /* Create a device context to draw in the overlay buffer with GDI. */
   MbufControl(MilOverlayImage, M_DC_ALLOC, M_DEFAULT);

   /* Inquire the device context. */
   hCustomDC = ((HDC)MbufInquire(MilOverlayImage, M_DC_HANDLE, M_NULL));

   /* Re-enable error printing */
   MappControl(M_DEFAULT, M_ERROR, M_PRINT_ENABLE);

   /* Perform operation if GDI drawing is supported. */
   if (hCustomDC)
      {
      POINT Hor[2];
      POINT Ver[2];
      SIZE  TxtSz;
      RECT  Txt;

      /* Draw a blue cross. */
      hpen=CreatePen(PS_SOLID, 1, RGB(0, 0, 255));
      hpenOld = SelectObject(hCustomDC, hpen);

      Hor[0].x = (MIL_INT32)0;
      Hor[0].y = (MIL_INT32)(ImageHeight/2);
      Hor[1].x = (MIL_INT32)ImageWidth;
      Hor[1].y = (MIL_INT32)(ImageHeight/2);
      Polyline(hCustomDC, Hor,2);

      Ver[0].x = (MIL_INT32)(ImageWidth/2);
      Ver[0].y = (MIL_INT32)0;
      Ver[1].x = (MIL_INT32)(ImageWidth/2);
      Ver[1].y = (MIL_INT32)ImageHeight;
      Polyline(hCustomDC, Ver,2);

      SelectObject(hCustomDC, hpenOld);
      DeleteObject(hpen);

      /* Prepare transparent text annotations. */
      SetBkMode(hCustomDC, TRANSPARENT);
      MosStrcpy(chText, 80, MIL_TEXT("GDI Overlay Text"));
      Count = (int) MosStrlen(chText);
      GetTextExtentPoint(hCustomDC, chText, Count, &TxtSz);


      /* Write red text. */
      Txt.left = (MIL_INT32)(ImageWidth*3/18);
      Txt.top  = (MIL_INT32)(ImageHeight*17/24);
      Txt.right  = (MIL_INT32)(Txt.left + TxtSz.cx);
      Txt.bottom = (MIL_INT32)(Txt.top  + TxtSz.cy);
      SetTextColor(hCustomDC,RGB(255, 0, 0));
      DrawText(hCustomDC, chText, Count, &Txt, DT_RIGHT);

      /* Write yellow text. */
      Txt.left = (MIL_INT32)ImageWidth*12/18;
      Txt.top  = (MIL_INT32)ImageHeight*17/24;
      Txt.right  = (MIL_INT32)(Txt.left + TxtSz.cx);
      Txt.bottom = (MIL_INT32)(Txt.top  + TxtSz.cy);
      SetTextColor(hCustomDC, RGB(255, 255, 0));
      DrawText(hCustomDC, chText, Count, &Txt, DT_RIGHT);

      /* Delete device context. */
      MbufControl(MilOverlayImage, M_DC_FREE, M_DEFAULT);

      /* Signal MIL that the overlay buffer was modified. */
      MbufControl(MilOverlayImage, M_MODIFIED, M_DEFAULT);
      }
#else
   /* Create a cairo surface to draw in the overlay buffer. */
   MbufControl(MilOverlayImage, M_SURFACE_ALLOC, M_COMPENSATION_ENABLE);

   /* Inquire the cairo surface. */
   cairo_surface_t * surface = ((cairo_surface_t *)MbufInquire(MilOverlayImage, M_SURFACE_HANDLE, M_NULL));
   if(surface)
      {
      cairo_t *cr;
      cr = cairo_create (surface);

      cairo_set_source_rgb (cr, 0, 0, 1);
      // Draw a blue cross in the overlay buffer.
      cairo_move_to (cr, 0, ImageHeight/2);
      cairo_line_to (cr, ImageWidth, ImageHeight/2);
      cairo_stroke (cr);
      cairo_move_to (cr, ImageWidth/2, 0);
      cairo_line_to (cr, ImageWidth/2, ImageHeight);
      cairo_stroke (cr);

      // Write Red text in the overlay buffer.
      MosStrcpy(chText, 80, "Cairo Overlay Text ");
      cairo_set_source_rgb (cr, 1, 0, 0);
      cairo_set_font_size(cr, 13);
      cairo_move_to(cr, ImageWidth*3/18, ImageHeight*4/6);
      cairo_show_text(cr, chText);

      // Write Yellow text in the overlay buffer.
      cairo_set_source_rgb (cr, 1, 1, 0);
      cairo_set_font_size(cr, 13);
      cairo_move_to(cr,  ImageWidth*12/18,ImageHeight*4/6);
      cairo_show_text(cr, chText);

      cairo_surface_flush(surface);
      cairo_destroy(cr);

      /* Delete device context. */
      MbufControl(MilOverlayImage, M_SURFACE_FREE, M_DEFAULT);

      /* Signal MIL that the overlay buffer was modified. */
      MbufControl(MilOverlayImage, M_MODIFIED, M_DEFAULT);
      }
#endif
    }

File: Matrox Imaging Library (MIL) X 23H2 (10.60)/Examples/General/MdispOverlay/Python/MdispOverlay.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
##########################################################################
#
# File name: MDispOverlay.py
#
# Synopsis:  This program shows how to display an image while creating
#            text and graphics annotations on top of it using MIL
#            graphic functions and windows GDI drawing under Windows.
#            If the target system supports grabbing, the annotations
#            are done on top of a continuous grab.
#
# Copyright © Matrox Electronic Systems Ltd., 1992-2023.
# All Rights Reserved
##########################################################################
 
import mil as MIL

# Target image.
IMAGE_FILE   = MIL.M_IMAGE_PATH + "Board.mim"

# Title for the display window.
WINDOW_TITLE = "Custom Title"

def MdispOverlayExample():

   # Allocate defaults.
   MilApplication, MilSystem, MilDisplay = MIL.MappAllocDefault(MIL.M_DEFAULT, DigIdPtr=MIL.M_NULL, ImageBufIdPtr=MIL.M_NULL)
   MilDigitizer = None
   # If the system has a digitizer, use it.
   if MIL.MsysInquire(MilSystem, MIL.M_DIGITIZER_NUM):
      MilDigitizer = MIL.MdigAlloc(MilSystem, MIL.M_DEFAULT, "M_DEFAULT", MIL.M_DEFAULT)
      MilImage = MIL.MbufAllocColor(MilSystem,
                                    MIL.MdigInquire(MilDigitizer, MIL.M_SIZE_BAND),
                                    MIL.MdigInquire(MilDigitizer, MIL.M_SIZE_X),
                                    MIL.MdigInquire(MilDigitizer, MIL.M_SIZE_Y),
                                    8 + MIL.M_UNSIGNED,
                                    MIL.M_IMAGE + MIL.M_DISP + MIL.M_PROC + MIL.M_GRAB)
      MIL.MbufClear(MilImage, 0)
   
   else:
      # Restore a static image.
      MilImage = MIL.MbufRestore(IMAGE_FILE, MilSystem)

   # Change display window title.
   MIL.MdispControl(MilDisplay, MIL.M_TITLE, WINDOW_TITLE)

   # Display the image buffer.
   MIL.MdispSelect(MilDisplay, MilImage)

   # Draw text and graphis annotations in the display overlay.
   OverlayDraw(MilDisplay)

   # If the system supports it, grab continuously in the displayed image. 
   if (MilDigitizer):
      MIL.MdigGrabContinuous(MilDigitizer, MilImage)

   # Pause to show the image.
   print("\nOVERLAY ANNOTATIONS:")
   print("--------------------\n")
   print("Displaying an image with overlay annotations.")
   print("Press <Enter> to continue.\n")
   MIL.MosGetch()

   # Stop the continuous grab and free digitizer if needed.
   if (MilDigitizer):
      MIL.MdigHalt(MilDigitizer)
      MIL.MdigFree(MilDigitizer)

      # Pause to show the result. 
      print("Displaying the last grabbed image.")
      print("Press <Enter> to end.\n")
      MIL.MosGetch()

   # Free image.
   MIL.MbufFree(MilImage)   

   # Free default allocations.
   MIL.MappFreeDefault(MilApplication, MilSystem, MilDisplay, MIL.M_NULL, MIL.M_NULL)

   return 0

# --------------------------------------------------------------- 
# Name:      OverlayDraw
# Synopsis:  This function draws annotations in the display overlay.

def OverlayDraw(MilDisplay):

   # Prepare overlay buffer. #
   ###########################

   # Enable the display of the overlay annotations.
   MIL.MdispControl(MilDisplay, MIL.M_OVERLAY, MIL.M_ENABLE)

   # Inquire the overlay buffer associated with the display. 
   MilOverlayImage = MIL.MdispInquire(MilDisplay, MIL.M_OVERLAY_ID)

   # Clear the overlay to transparent. 
   MIL.MdispControl(MilDisplay, MIL.M_OVERLAY_CLEAR, MIL.M_DEFAULT)

   # Disable the overlay display update to accelerate annotations. 
   MIL.MdispControl(MilDisplay, MIL.M_OVERLAY_SHOW, MIL.M_DISABLE)

   # Inquire overlay size. 
   ImageWidth  = MIL.MbufInquire(MilOverlayImage, MIL.M_SIZE_X)
   ImageHeight = MIL.MbufInquire(MilOverlayImage, MIL.M_SIZE_Y)

   # Draw MIL overlay annotations. #
   #################################

   # Set the graphic text background to transparent. 
   MIL.MgraControl(MIL.M_DEFAULT, MIL.M_BACKGROUND_MODE, MIL.M_TRANSPARENT)

   # Print a white string in the overlay image buffer. 
   MIL.MgraColor(MIL.M_DEFAULT, MIL.M_COLOR_WHITE)
   MIL.MgraText(MIL.M_DEFAULT, MilOverlayImage, ImageWidth/9, ImageHeight/5,    " -------------------- ")
   MIL.MgraText(MIL.M_DEFAULT, MilOverlayImage, ImageWidth/9, ImageHeight/5+25, " - MIL Overlay Text - ")
   MIL.MgraText(MIL.M_DEFAULT, MilOverlayImage, ImageWidth/9, ImageHeight/5+50, " -------------------- ")

   # Print a green string in the overlay image buffer. 
   MIL.MgraColor(MIL.M_DEFAULT, MIL.M_COLOR_GREEN)
   MIL.MgraText(MIL.M_DEFAULT, MilOverlayImage, ImageWidth*11/18, ImageHeight/5,    " ---------------------")
   MIL.MgraText(MIL.M_DEFAULT, MilOverlayImage, ImageWidth*11/18, ImageHeight/5+25, " - MIL Overlay Text - ")
   MIL.MgraText(MIL.M_DEFAULT, MilOverlayImage, ImageWidth*11/18, ImageHeight/5+50, " ---------------------")

   # Re-enable the overlay display after all annotations are done. 
   MIL.MdispControl(MilDisplay, MIL.M_OVERLAY_SHOW, MIL.M_ENABLE)

   try:
      import win32ui
      import win32api
      import win32con

      # Draw GDI color overlay annotation #
      #####################################

      # Disable error printing for the inquire might not be supported.
      MIL.MappControl(MIL.M_DEFAULT, MIL.M_ERROR, MIL.M_PRINT_DISABLE)

      # Create a device context to draw in the overlay buffer with GDI. 
      MIL.MbufControl(MilOverlayImage, MIL.M_DC_ALLOC, MIL.M_DEFAULT)
      # Inquire the device context. 
      hCustomDC = win32ui.CreateDCFromHandle(MIL.MbufInquire(MilOverlayImage, MIL.M_DC_HANDLE))

      # Re-enable error printing. 
      MIL.MappControl(MIL.M_DEFAULT, MIL.M_ERROR, MIL.M_PRINT_ENABLE)

      # Perform operation if GDI drawing is supported. 
      if hCustomDC:
         Hor = [(0, int(ImageHeight / 2)), (int(ImageWidth), int(ImageHeight / 2))]
         Ver = [(int(ImageWidth / 2), 0), (int(ImageWidth / 2), int(ImageHeight))]

         # Draw a blue cross.
         hpen = win32ui.CreatePen(win32con.PS_SOLID, 1, win32api.RGB(0, 0, 255))
         hpenOld = hCustomDC.SelectObject(hpen)

         hCustomDC.Polyline(Hor)
         hCustomDC.Polyline(Ver)

         hCustomDC.SelectObject(hpenOld)

         # Prepare transparent text annotations. 
         hCustomDC.SetBkMode(win32con.TRANSPARENT)
         chText = "GDI Overlay Text"
         (x, y) = hCustomDC.GetTextExtentPoint(chText)

         # Write red text.
         Txt = (int(ImageWidth*3/18), int(ImageHeight*17/24), int(ImageWidth*3/18) + x, int(ImageHeight*17/24) + y)
         hCustomDC.SetTextColor(win32api.RGB(255, 0, 0))
         hCustomDC.DrawText(chText, Txt, win32con.DT_RIGHT)

         # Write yellow text.
         Txt = (int(ImageWidth*12/18), int(ImageHeight*17/24), int(ImageWidth*12/18) + x, int(ImageHeight*17/24) + y)
         hCustomDC.SetTextColor(win32api.RGB(255, 255, 0))
         hCustomDC.DrawText(chText, Txt, win32con.DT_RIGHT)

         # Delete device context. 
         MIL.MbufControl(MilOverlayImage, MIL.M_DC_FREE, MIL.M_DEFAULT)

         # Signal MIL that the overlay buffer was modified. 
         MIL.MbufControl(MilOverlayImage, MIL.M_MODIFIED, MIL.M_DEFAULT)
   except:
      pass
      # GDI drawing is not supported

if __name__ == "__main__":
   MdispOverlayExample()
